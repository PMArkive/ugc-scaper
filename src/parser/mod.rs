use crate::{ParseError, Result};
use scraper::{ElementRef, Selector};
use steamid_ng::SteamID;
use time::format_description::FormatItem;
use time::macros::format_description;

mod player;
mod player_details;
mod seasons;
mod team;
mod team_lookup;
mod team_matches;
mod team_roster_history;

pub use player::*;
pub use player_details::*;
pub use seasons::*;
pub use team::*;
pub use team_lookup::*;
pub use team_matches::*;
pub use team_roster_history::*;

pub trait Parser {
    type Output;
    fn parse(&self, document: &str) -> Result<Self::Output>;
}

trait ElementExt<'a> {
    fn first_text(&'a self) -> Option<&'a str>;
    fn nth_text(&'a self, n: usize) -> Option<&'a str>;
}

impl<'a> ElementExt<'a> for ElementRef<'a> {
    fn first_text(&self) -> Option<&'a str> {
        self.text().map(str::trim).find(|s| !s.is_empty())
    }
    fn nth_text(&self, n: usize) -> Option<&'a str> {
        self.text()
            .filter(|s| !s.trim().is_empty())
            .nth(n - 1)
            .map(str::trim)
    }
}

fn select_text<'a>(el: ElementRef<'a>, selector: &Selector) -> Option<&'a str> {
    el.select(selector)
        .next()
        .and_then(|item| item.text().find(|s| !s.trim().is_empty()))
        .map(str::trim)
}

fn select_last_text<'a>(el: ElementRef<'a>, selector: &Selector) -> Option<&'a str> {
    el.select(selector)
        .next()
        .and_then(|item| item.text().last())
        .map(str::trim)
}

const DATE_FORMAT: &[FormatItem<'static>] =
    format_description!("[month padding:none]/[day padding:none]/[year]");
const MEMBER_DATE_FORMAT: &[FormatItem<'static>] = format_description!(
    "[month repr:short] [day padding:none], [year]\n/\n[hour padding:none]:[minute] [period]\n(ET)"
);
const ROSTER_HISTORY_DATE_FORMAT: &[FormatItem<'static>] =
    format_description!("[month repr:short] [day padding:none], [year]");

fn team_id_from_link(link: &str) -> Result<u32, ParseError> {
    link.rsplit_once('=')
        .and_then(|part| part.1.parse().ok())
        .ok_or_else(|| ParseError::InvalidLink {
            link: link.to_string(),
            role: "team id",
        })
}

fn steam_id_from_link(link: &str) -> Result<SteamID, ParseError> {
    link.rsplit_once('=')
        .and_then(|part| part.1.parse::<u64>().ok())
        .ok_or_else(|| ParseError::InvalidLink {
            link: link.to_string(),
            role: "user id",
        })
        .map(SteamID::from)
}
