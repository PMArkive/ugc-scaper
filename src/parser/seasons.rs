use super::Parser;
use crate::data::{Season, Seasons};
use crate::parser::{select_text, ElementExt};
use crate::{ParseError, Result};
use scraper::{Html, Selector};

const SELECTOR_MENU: &str = ".sub-menu";
const SELECTOR_NAME: &str = ".mega-menu-sub-title";
const SELECTOR_SEASON_LINK: &str = "ul[id$=\"seasons\"] a[href^=\"rankings_\"]";

pub struct SeasonsParser {
    selector_menu: Selector,
    selector_name: Selector,
    selector_link: Selector,
}

impl Default for SeasonsParser {
    fn default() -> Self {
        Self::new()
    }
}

impl SeasonsParser {
    pub fn new() -> Self {
        SeasonsParser {
            selector_menu: Selector::parse(SELECTOR_MENU).unwrap(),
            selector_name: Selector::parse(SELECTOR_NAME).unwrap(),
            selector_link: Selector::parse(SELECTOR_SEASON_LINK).unwrap(),
        }
    }
}

impl Parser for SeasonsParser {
    type Output = Vec<Seasons>;

    fn parse(&self, document: &str) -> Result<Self::Output> {
        let document = Html::parse_document(document);

        document
            .select(&self.selector_menu)
            .filter(|item| item.select(&self.selector_link).next().is_some())
            .filter(|item| item.select(&self.selector_name).next().is_some())
            .map(|item| {
                let name = select_text(item, &self.selector_name)
                    .ok_or(ParseError::EmptyText {
                        role: "game mode name",
                        selector: SELECTOR_NAME,
                    })?
                    .trim_end_matches(" Menu");

                let seasons = item
                    .select(&self.selector_link)
                    .map(|link| {
                        let text = link
                            .first_text()
                            .ok_or(ParseError::EmptyText {
                                role: "season name",
                                selector: SELECTOR_SEASON_LINK,
                            })?
                            .trim_end_matches(" Final Standings")
                            .trim_end_matches(" Final Rank")
                            .trim_end_matches(" Final Ranks");
                        let link = link.attr("href").ok_or(ParseError::EmptyText {
                            role: "season link",
                            selector: SELECTOR_SEASON_LINK,
                        })?;
                        let id = link
                            .trim_start_matches("rankings_")
                            .trim_end_matches(".cfm");
                        Ok(Season {
                            name: text.to_string(),
                            id: id.to_string(),
                        })
                    })
                    .collect::<Result<_>>()?;

                Ok(Seasons {
                    mode: name.to_string(),
                    seasons,
                })
            })
            .collect::<Result<Vec<_>>>()
    }
}
