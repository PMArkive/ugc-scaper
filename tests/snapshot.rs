#[cfg(feature = "serde")]
use insta::assert_json_snapshot;
#[cfg(feature = "serde")]
use std::fs::read_to_string;
#[cfg(feature = "serde")]
use test_case::test_case;
#[cfg(feature = "serde")]
use ugc_scraper::parser::{
    MapHistoryParser, MatchPageParser, Parser, PlayerDetailsParser, PlayerParser, SeasonsParser,
    TeamLookupParser, TeamMatchesParser, TeamParser, TeamRosterHistoryParser, TransactionParser,
};

#[test_case("player_76561198024494988.html", "player")]
#[test_case("player_76561198049312442.html", "player_classes")]
#[cfg(feature = "serde")]
fn test_parse_player_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = PlayerParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("player_details_76561198024494988.html", "player_details")]
#[cfg(feature = "serde")]
fn test_parse_player_details_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = PlayerDetailsParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("team_7861.html", "team")]
#[test_case("team_4105.html", "older_team")]
#[test_case("team_8157.html", "team_no_tz")]
#[test_case("team_6929.html", "team_changed_name")]
#[test_case("team_32437.html", "team_empty_name_change")]
#[test_case("team_29228.html", "team_newlines_join_date")]
#[test_case("team_10763.html", "team_na_4v4")]
#[cfg(feature = "serde")]
fn test_parse_team_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("team_roster_history_7861.html", "team_roster_history")]
#[test_case("team_roster_history_2133.html", "team_roster_history_no_group")]
#[cfg(feature = "serde")]
fn test_parse_team_roster_history_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamRosterHistoryParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("team_matches_7861.html", "team_matches")]
#[cfg(feature = "serde")]
fn test_parse_team_matches_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamMatchesParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("index.html", "seasons")]
#[cfg(feature = "serde")]
fn test_parse_seasons_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = SeasonsParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("lookup_9v9.html", "seasons_9")]
#[test_case("lookup_6v6.html", "seasons_6")]
#[test_case("lookup_4v4.html", "seasons_4")]
#[test_case("lookup_2v2.html", "seasons_2")]
#[cfg(feature = "serde")]
fn test_parse_seasons_mode_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TeamLookupParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("match_116246.html", "match")]
#[cfg(feature = "serde")]
fn test_parse_match_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = MatchPageParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("transactions_4v4.html", "transaction")]
#[cfg(feature = "serde")]
fn test_parse_transaction_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = TransactionParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}

#[test_case("map_9v9.html", "maps_9")]
#[test_case("map_9v9_2.html", "maps_9_2")]
#[test_case("map_6v6.html", "maps_6")]
#[test_case("map_4v4.html", "maps_4")]
#[test_case("map_2v2.html", "maps_2")]
#[cfg(feature = "serde")]
fn test_parse_maps_html(input: &str, name: &str) {
    let body = read_to_string(format!("tests/data/{input}")).unwrap();
    let parser = MapHistoryParser::new();
    let parsed = parser.parse(&body).unwrap();
    assert_json_snapshot!(format!("parse_{name}_html"), parsed);
}
