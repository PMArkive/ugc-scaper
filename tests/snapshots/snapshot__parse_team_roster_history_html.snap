---
source: tests/snapshot.rs
expression: parsed
---
[
  {
    "name": "marko",
    "steam_id": 76561198274165935,
    "joined": "2023-10-19",
    "left": null
  },
  {
    "name": "NioH",
    "steam_id": 76561198218881647,
    "joined": "2023-07-22",
    "left": null
  },
  {
    "name": "Teroantero2007",
    "steam_id": 76561197996902035,
    "joined": "2023-06-26",
    "left": null
  },
  {
    "name": "Raipe",
    "steam_id": 76561198061082936,
    "joined": "2023-03-19",
    "left": null
  },
  {
    "name": "drew",
    "steam_id": 76561198012304706,
    "joined": "2023-02-12",
    "left": null
  },
  {
    "name": "BaaBo",
    "steam_id": 76561198004331478,
    "joined": "2023-01-09",
    "left": null
  },
  {
    "name": "bat",
    "steam_id": 76561198019511460,
    "joined": "2022-10-08",
    "left": "2023-03-19"
  },
  {
    "name": "Houf",
    "steam_id": 76561198448050535,
    "joined": "2022-09-26",
    "left": "2022-10-05"
  },
  {
    "name": "Royal Flush",
    "steam_id": 76561198052084714,
    "joined": "2022-05-08",
    "left": null
  },
  {
    "name": "Kireek",
    "steam_id": 76561198052694464,
    "joined": "2022-01-27",
    "left": null
  },
  {
    "name": "Husky",
    "steam_id": 76561198176360750,
    "joined": "2021-11-07",
    "left": "2022-05-15"
  },
  {
    "name": "Croc",
    "steam_id": 76561198902479382,
    "joined": "2021-10-04",
    "left": "2022-05-02"
  },
  {
    "name": "Piney",
    "steam_id": 76561198050910939,
    "joined": "2021-10-04",
    "left": "2023-08-04"
  },
  {
    "name": "Airplay",
    "steam_id": 76561198286047751,
    "joined": "2021-10-04",
    "left": "2023-06-26"
  },
  {
    "name": "kanon",
    "steam_id": 76561198154189208,
    "joined": "2021-09-29",
    "left": "2022-01-04"
  },
  {
    "name": "Husky",
    "steam_id": 76561198176360750,
    "joined": "2021-07-06",
    "left": "2021-10-03"
  },
  {
    "name": "schaffi",
    "steam_id": 76561198061369274,
    "joined": "2021-06-12",
    "left": "2023-07-22"
  },
  {
    "name": "Solar",
    "steam_id": 76561198015394193,
    "joined": "2021-06-10",
    "left": "2021-09-17"
  },
  {
    "name": "Spork",
    "steam_id": 76561198208448312,
    "joined": "2021-06-09",
    "left": "2022-09-30"
  },
  {
    "name": "Royal Flush",
    "steam_id": 76561198052084714,
    "joined": "2021-06-09",
    "left": "2022-01-31"
  },
  {
    "name": "Derakusa",
    "steam_id": 76561198011495003,
    "joined": "2020-10-01",
    "left": null
  },
  {
    "name": "Schaffi",
    "steam_id": 76561198061369274,
    "joined": "2020-10-01",
    "left": "2021-02-15"
  },
  {
    "name": "Bonobo",
    "steam_id": 76561198038822935,
    "joined": "2020-06-07",
    "left": "2021-07-06"
  },
  {
    "name": "Dempsey",
    "steam_id": 76561198049405872,
    "joined": "2020-06-04",
    "left": "2021-11-07"
  },
  {
    "name": "clx",
    "steam_id": 76561198310925535,
    "joined": "2020-05-20",
    "left": "2020-05-25"
  },
  {
    "name": "viking",
    "steam_id": 76561198162124761,
    "joined": "2020-02-16",
    "left": "2021-10-04"
  },
  {
    "name": "Ikaros",
    "steam_id": 76561198158482651,
    "joined": "2019-11-03",
    "left": "2021-10-04"
  },
  {
    "name": "StelaZ",
    "steam_id": 76561198036824480,
    "joined": "2019-09-29",
    "left": null
  },
  {
    "name": "dt",
    "steam_id": 76561198120760515,
    "joined": "2019-09-22",
    "left": "2021-02-15"
  },
  {
    "name": "brandon",
    "steam_id": 76561198073579100,
    "joined": "2019-09-12",
    "left": "2020-03-10"
  },
  {
    "name": "Derakusa",
    "steam_id": 76561198011495003,
    "joined": "2019-06-21",
    "left": "2020-09-21"
  },
  {
    "name": "WiseGenie",
    "steam_id": 76561198405968967,
    "joined": "2019-06-05",
    "left": "2021-03-01"
  },
  {
    "name": "Amadeus",
    "steam_id": 76561198052198679,
    "joined": "2019-05-18",
    "left": "2020-02-16"
  },
  {
    "name": "Sebbo",
    "steam_id": 76561198054395451,
    "joined": "2019-02-23",
    "left": "2019-09-30"
  },
  {
    "name": "Scrambled",
    "steam_id": 76561198058283562,
    "joined": "2019-02-20",
    "left": "2019-09-12"
  },
  {
    "name": "Peace Sells",
    "steam_id": 76561198224755980,
    "joined": "2019-02-03",
    "left": "2020-10-01"
  },
  {
    "name": "Slappy",
    "steam_id": 76561198017238822,
    "joined": "2019-01-20",
    "left": "2019-06-21"
  },
  {
    "name": "Polarium",
    "steam_id": 76561198208448312,
    "joined": "2019-01-16",
    "left": "2019-05-03"
  },
  {
    "name": "Chochy",
    "steam_id": 76561198275783619,
    "joined": "2019-01-14",
    "left": "2019-01-29"
  },
  {
    "name": "Kaga",
    "steam_id": 76561198040965137,
    "joined": "2018-11-08",
    "left": null
  },
  {
    "name": "Quaddeln",
    "steam_id": 76561198079948125,
    "joined": "2018-10-20",
    "left": "2019-09-29"
  },
  {
    "name": "supra",
    "steam_id": 76561198011558250,
    "joined": "2018-10-03",
    "left": "2018-12-13"
  },
  {
    "name": "GibZ",
    "steam_id": 76561198027463207,
    "joined": "2018-09-27",
    "left": "2019-01-17"
  },
  {
    "name": "Pine",
    "steam_id": 76561198050910939,
    "joined": "2018-09-20",
    "left": "2021-01-25"
  },
  {
    "name": "Polarium",
    "steam_id": 76561198208448312,
    "joined": "2018-09-06",
    "left": "2018-11-29"
  },
  {
    "name": "Mr British",
    "steam_id": 76561198046899720,
    "joined": "2018-05-27",
    "left": "2018-10-28"
  },
  {
    "name": "MrTK",
    "steam_id": 76561198049707157,
    "joined": "2018-05-21",
    "left": "2019-02-20"
  },
  {
    "name": "Pretzel",
    "steam_id": 76561197996215628,
    "joined": "2018-05-18",
    "left": "2018-10-20"
  },
  {
    "name": "Teeru",
    "steam_id": 76561198148148180,
    "joined": "2018-05-15",
    "left": "2018-09-11"
  },
  {
    "name": "Deepfried Monk",
    "steam_id": 76561198028288946,
    "joined": "2018-05-15",
    "left": "2020-06-07"
  },
  {
    "name": "Leonio",
    "steam_id": 76561198078395610,
    "joined": "2018-05-15",
    "left": "2019-09-22"
  },
  {
    "name": "Vinegar",
    "steam_id": 76561197983224631,
    "joined": "2018-02-26",
    "left": "2018-04-29"
  },
  {
    "name": "disabled caiz",
    "steam_id": 76561198082325306,
    "joined": "2018-01-29",
    "left": "2018-10-03"
  },
  {
    "name": "Mankind LFT",
    "steam_id": 76561197999362239,
    "joined": "2017-12-26",
    "left": "2018-01-22"
  },
  {
    "name": "Tatty G",
    "steam_id": 76561198091359665,
    "joined": "2017-10-21",
    "left": "2018-09-20"
  },
  {
    "name": "Rem",
    "steam_id": 76561198065231718,
    "joined": "2017-10-21",
    "left": "2018-07-04"
  },
  {
    "name": "Mr British 3",
    "steam_id": 76561198046899720,
    "joined": "2017-10-13",
    "left": "2018-03-22"
  },
  {
    "name": "Polarium",
    "steam_id": 76561198208448312,
    "joined": "2017-10-05",
    "left": "2017-11-28"
  },
  {
    "name": "PlayerProDG",
    "steam_id": 76561198087728099,
    "joined": "2017-06-03",
    "left": "2018-04-21"
  },
  {
    "name": "Royal Flush",
    "steam_id": 76561198052084714,
    "joined": "2017-05-22",
    "left": "2019-02-23"
  },
  {
    "name": "Vinegar TheWon",
    "steam_id": 76561197983224631,
    "joined": "2017-05-14",
    "left": "2018-01-05"
  },
  {
    "name": "HiipFire",
    "steam_id": 76561198012785231,
    "joined": "2017-05-14",
    "left": "2017-05-26"
  },
  {
    "name": "coca",
    "steam_id": 76561198077605729,
    "joined": "2017-04-20",
    "left": "2017-08-23"
  },
  {
    "name": "Vouri",
    "steam_id": 76561198106656127,
    "joined": "2017-02-19",
    "left": "2017-12-29"
  },
  {
    "name": "Jedi(fla)maste",
    "steam_id": 76561197990486664,
    "joined": "2017-01-17",
    "left": null
  },
  {
    "name": "Teroantero",
    "steam_id": 76561197996902035,
    "joined": "2017-01-17",
    "left": "2017-05-16"
  },
  {
    "name": "TC Coca",
    "steam_id": 76561198077605729,
    "joined": "2017-01-09",
    "left": "2017-01-20"
  },
  {
    "name": "Pendulum",
    "steam_id": 76561198103892101,
    "joined": "2016-12-30",
    "left": "2018-05-14"
  },
  {
    "name": "Strix",
    "steam_id": 76561198078178141,
    "joined": "2016-12-17",
    "left": "2016-12-22"
  },
  {
    "name": "Deepfried Monk",
    "steam_id": 76561198028288946,
    "joined": "2016-09-26",
    "left": "2016-12-30"
  },
  {
    "name": "bfisch",
    "steam_id": 76561198052461274,
    "joined": "2016-09-25",
    "left": "2017-01-06"
  },
  {
    "name": "SF Iro",
    "steam_id": 76561198025643013,
    "joined": "2016-09-22",
    "left": "2016-12-27"
  },
  {
    "name": "meZz09",
    "steam_id": 76561198059887260,
    "joined": "2016-09-15",
    "left": "2016-12-22"
  },
  {
    "name": "denkra",
    "steam_id": 76561198074899144,
    "joined": "2016-09-13",
    "left": "2016-10-04"
  },
  {
    "name": "Capu",
    "steam_id": 76561198044857348,
    "joined": "2016-06-26",
    "left": "2016-09-26"
  },
  {
    "name": "GIELE JohsKo2",
    "steam_id": 76561198058301893,
    "joined": "2016-06-21",
    "left": "2017-01-12"
  },
  {
    "name": "Kimo",
    "steam_id": 76561197992327511,
    "joined": "2016-06-19",
    "left": null
  },
  {
    "name": "Deepfried Monk",
    "steam_id": 76561198028288946,
    "joined": "2016-05-29",
    "left": "2016-09-15"
  },
  {
    "name": "Foz",
    "steam_id": 76561198076014163,
    "joined": "2016-05-25",
    "left": null
  },
  {
    "name": "Cmm Bert",
    "steam_id": 76561198016658674,
    "joined": "2016-05-23",
    "left": "2017-10-21"
  },
  {
    "name": "Vinegar",
    "steam_id": 76561197983224631,
    "joined": "2016-05-15",
    "left": "2016-06-12"
  },
  {
    "name": "Kireek",
    "steam_id": 76561198052694464,
    "joined": "2016-05-12",
    "left": "2019-05-13"
  },
  {
    "name": "Ghillie",
    "steam_id": 76561198033193346,
    "joined": "2016-02-26",
    "left": "2016-09-18"
  },
  {
    "name": "jojo",
    "steam_id": 76561197995029224,
    "joined": "2016-01-17",
    "left": null
  },
  {
    "name": "BaaBo",
    "steam_id": 76561198004331478,
    "joined": "2016-01-15",
    "left": "2017-03-10"
  },
  {
    "name": "Deity",
    "steam_id": 76561198076020012,
    "joined": "2015-12-29",
    "left": null
  },
  {
    "name": "Bencer",
    "steam_id": 76561198042734399,
    "joined": "2015-11-05",
    "left": "2015-12-17"
  },
  {
    "name": "froggie",
    "steam_id": 76561197980091914,
    "joined": "2015-10-15",
    "left": "2015-12-12"
  },
  {
    "name": "Mak",
    "steam_id": 76561198049312442,
    "joined": "2015-09-23",
    "left": null
  },
  {
    "name": "Vouri",
    "steam_id": 76561198106656127,
    "joined": "2015-09-22",
    "left": "2016-09-21"
  },
  {
    "name": "cheesey",
    "steam_id": 76561198058352657,
    "joined": "2015-09-21",
    "left": "2017-10-21"
  },
  {
    "name": "Teo",
    "steam_id": 76561198030853412,
    "joined": "2015-09-21",
    "left": "2016-05-23"
  },
  {
    "name": "corr",
    "steam_id": 76561198046921199,
    "joined": "2015-07-13",
    "left": "2015-09-02"
  },
  {
    "name": "Strno",
    "steam_id": 76561198015251809,
    "joined": "2015-07-05",
    "left": "2016-12-04"
  },
  {
    "name": "cobz",
    "steam_id": 76561198016371188,
    "joined": "2015-06-28",
    "left": "2015-11-05"
  },
  {
    "name": "Fourty",
    "steam_id": 76561197970532785,
    "joined": "2015-06-26",
    "left": "2016-06-18"
  },
  {
    "name": "NightWing",
    "steam_id": 76561198010561736,
    "joined": "2015-06-25",
    "left": "2015-09-22"
  },
  {
    "name": "Feyne",
    "steam_id": 76561198028389239,
    "joined": "2015-06-24",
    "left": "2016-05-25"
  },
  {
    "name": "DereK",
    "steam_id": 76561198072932719,
    "joined": "2015-05-31",
    "left": "2015-09-17"
  },
  {
    "name": "Rem",
    "steam_id": 76561198065231718,
    "joined": "2015-05-28",
    "left": "2016-01-01"
  },
  {
    "name": "Capu",
    "steam_id": 76561198044857348,
    "joined": "2015-01-26",
    "left": "2015-05-05"
  },
  {
    "name": "Janqe",
    "steam_id": 76561198025732057,
    "joined": "2015-01-26",
    "left": "2015-07-13"
  },
  {
    "name": "THE METHVAN",
    "steam_id": 76561198130935924,
    "joined": "2015-01-25",
    "left": "2015-05-01"
  },
  {
    "name": "Panther",
    "steam_id": 76561198062889828,
    "joined": "2015-01-24",
    "left": "2015-05-25"
  },
  {
    "name": "Slappy",
    "steam_id": 76561198017238822,
    "joined": "2015-01-23",
    "left": "2017-09-10"
  },
  {
    "name": "Gatinmnsok",
    "steam_id": 76561198064715318,
    "joined": "2015-01-01",
    "left": "2016-06-26"
  },
  {
    "name": "Rosstisfer",
    "steam_id": 76561198019933879,
    "joined": "2014-11-04",
    "left": "2016-09-10"
  },
  {
    "name": "Darkes",
    "steam_id": 76561198058253854,
    "joined": "2014-11-04",
    "left": "2014-12-26"
  },
  {
    "name": "khaos",
    "steam_id": 76561198044610420,
    "joined": "2014-10-13",
    "left": "2015-07-05"
  },
  {
    "name": "Super",
    "steam_id": 76561198038067254,
    "joined": "2014-10-07",
    "left": "2014-12-26"
  },
  {
    "name": "DeevoDwarf",
    "steam_id": 76561198058182064,
    "joined": "2014-09-22",
    "left": "2014-09-22"
  },
  {
    "name": "Derakusa",
    "steam_id": 76561198011495003,
    "joined": "2014-09-18",
    "left": "2015-09-23"
  },
  {
    "name": "Shoosh",
    "steam_id": 76561198049593717,
    "joined": "2014-09-12",
    "left": null
  },
  {
    "name": "Rem",
    "steam_id": 76561198065231718,
    "joined": "2014-08-17",
    "left": "2015-05-28"
  },
  {
    "name": "DemoN",
    "steam_id": 76561198059584839,
    "joined": "2014-06-21",
    "left": "2016-05-01"
  },
  {
    "name": "DevilHunter",
    "steam_id": 76561198049826730,
    "joined": "2014-06-21",
    "left": "2016-05-01"
  },
  {
    "name": "Arty",
    "steam_id": 76561198061685333,
    "joined": "2014-06-07",
    "left": "2014-08-29"
  },
  {
    "name": "Boskov",
    "steam_id": 76561198054280187,
    "joined": "2014-05-18",
    "left": "2014-08-20"
  },
  {
    "name": "TinyTerror",
    "steam_id": 76561197996307844,
    "joined": "2014-05-10",
    "left": "2016-01-30"
  },
  {
    "name": "hjarteknusar",
    "steam_id": 76561198023996045,
    "joined": "2014-04-30",
    "left": "2015-01-23"
  },
  {
    "name": "Fish",
    "steam_id": 76561198052362074,
    "joined": "2014-04-30",
    "left": null
  },
  {
    "name": "drtyeyeball",
    "steam_id": 76561198056603660,
    "joined": "2014-04-29",
    "left": "2015-01-22"
  },
  {
    "name": "Bencer",
    "steam_id": 76561198042734399,
    "joined": "2014-04-26",
    "left": "2014-05-27"
  },
  {
    "name": "Mircea YK",
    "steam_id": 76561198047488650,
    "joined": "2014-04-26",
    "left": "2014-07-31"
  },
  {
    "name": "smarties",
    "steam_id": 76561198054288197,
    "joined": "2014-03-09",
    "left": "2014-04-27"
  },
  {
    "name": "DevilHunter",
    "steam_id": 76561198049826730,
    "joined": "2014-03-05",
    "left": "2014-04-26"
  },
  {
    "name": "theshinyspy",
    "steam_id": 76561198067493714,
    "joined": "2014-03-05",
    "left": "2014-06-06"
  },
  {
    "name": "Rosstisfer",
    "steam_id": 76561198019933879,
    "joined": "2014-03-03",
    "left": "2014-09-11"
  },
  {
    "name": "Royal Nipple",
    "steam_id": 76561198072518628,
    "joined": "2014-02-22",
    "left": "2014-06-16"
  },
  {
    "name": "dr bad bone ty",
    "steam_id": 76561198052922917,
    "joined": "2014-02-13",
    "left": "2014-02-17"
  },
  {
    "name": "hjarteknusar",
    "steam_id": 76561198023996045,
    "joined": "2014-01-17",
    "left": "2014-03-11"
  },
  {
    "name": "Deepfried Monk",
    "steam_id": 76561198028288946,
    "joined": "2014-01-01",
    "left": "2014-01-17"
  },
  {
    "name": "Ghillie",
    "steam_id": 76561198033193346,
    "joined": "2014-01-01",
    "left": "2014-09-17"
  },
  {
    "name": "ekkelund",
    "steam_id": 76561197997105682,
    "joined": "2013-12-31",
    "left": "2014-02-03"
  },
  {
    "name": "-Xe- Vappppppp",
    "steam_id": 76561198019933879,
    "joined": "2013-12-27",
    "left": "2014-02-21"
  },
  {
    "name": "wardemon mix",
    "steam_id": 76561198054462519,
    "joined": "2013-12-27",
    "left": "2014-05-04"
  },
  {
    "name": "DudeoftheRock",
    "steam_id": 76561198059401865,
    "joined": "2013-09-28",
    "left": "2015-01-23"
  },
  {
    "name": "Spannzer",
    "steam_id": 76561198057121917,
    "joined": "2013-09-19",
    "left": "2013-10-21"
  },
  {
    "name": "Feyne",
    "steam_id": 76561198028389239,
    "joined": "2013-09-03",
    "left": "2015-01-23"
  },
  {
    "name": "Shoosh!",
    "steam_id": 76561198049593717,
    "joined": "2013-08-22",
    "left": "2014-05-21"
  },
  {
    "name": "Crazy3k",
    "steam_id": 76561198056225108,
    "joined": "2013-08-15",
    "left": "2014-04-28"
  },
  {
    "name": "Hayras",
    "steam_id": 76561198003255057,
    "joined": "2013-08-12",
    "left": "2013-08-22"
  },
  {
    "name": "Gardener Gnome",
    "steam_id": 76561198010266253,
    "joined": "2013-08-11",
    "left": "2015-01-23"
  },
  {
    "name": "HardCoreMother",
    "steam_id": 76561198021813586,
    "joined": "2013-08-11",
    "left": "2013-08-12"
  },
  {
    "name": "drtyeyeball_#0",
    "steam_id": 76561198056603660,
    "joined": "2013-08-09",
    "left": "2013-12-27"
  },
  {
    "name": "iTouch",
    "steam_id": 76561198041855608,
    "joined": "2013-08-09",
    "left": "2014-01-01"
  },
  {
    "name": "? Wolfy ?",
    "steam_id": 76561198072470593,
    "joined": "2013-08-09",
    "left": "2013-12-27"
  },
  {
    "name": "Icewind",
    "steam_id": 76561198024494988,
    "joined": "2013-08-09",
    "left": null
  },
  {
    "name": "Huper",
    "steam_id": 76561198007918526,
    "joined": "2013-08-06",
    "left": "2014-04-28"
  },
  {
    "name": "BenceMate",
    "steam_id": 76561198042734399,
    "joined": "2013-08-06",
    "left": "2014-01-01"
  },
  {
    "name": "ShakeNova",
    "steam_id": 76561198054279994,
    "joined": "2013-08-06",
    "left": "2015-01-23"
  },
  {
    "name": "jojo",
    "steam_id": 76561197995029224,
    "joined": "2013-08-06",
    "left": "2015-10-04"
  },
  {
    "name": "NoSocks",
    "steam_id": 76561198012110404,
    "joined": "2013-08-06",
    "left": null
  },
  {
    "name": "-Xe-Shoosh!",
    "steam_id": 76561198049593717,
    "joined": "2013-08-06",
    "left": "2013-08-22"
  },
  {
    "name": "-Xe-Mezzoo!",
    "steam_id": 76561198060782652,
    "joined": "2013-08-06",
    "left": "2013-12-27"
  },
  {
    "name": "-Xe-Vaporean",
    "steam_id": 76561198019933879,
    "joined": "2013-08-06",
    "left": "2013-12-27"
  },
  {
    "name": "Exe",
    "steam_id": 76561198053667422,
    "joined": "2013-08-06",
    "left": "2016-05-27"
  }
]
